third_party/libunwindstack/tests/PeCoffUnwindInfoUnwinderX86_64Test.cpp

```diff
@@ -0,0 +1,1000 @@
+
+    stack_pointer_ -= sizeof(uint64_t);
+    process_mem_fake_->SetData64(stack_pointer_, options.return_address);
+
+    std::vector<UnwindCode> unwind_codes;
+    AddPushedRegisters(pushed_registers, &unwind_codes);
+    AddStackAllocation(options.stack_allocation, &unwind_codes);
+    AddSavedRegisters(saved_registers, &unwind_codes);
+    AddSavedXmm128Registers(saved_xmm128_registers, &unwind_codes);
+    AddFramePointerRegisterOp(options, &unwind_codes);
+
+    // TODO: Chained info.
+    uint8_t flags = 0x00;
+    uint8_t version_and_flags = flags | 0x01;
+
+    new_unwind_info.version_and_flags = version_and_flags;
+    new_unwind_info.prolog_size = unwind_codes.size() + 1;
+
+    // Unwind codes need to be saved in the order that they will be processed by the unwinding code,
+    // which is reverse to the order of machine instructions (which we simulate above).
+    std::reverse(unwind_codes.begin(), unwind_codes.end());
+
+    CHECK(unwind_codes.size() <= std::numeric_limits<uint8_t>::max());
+    new_unwind_info.num_codes = static_cast<uint8_t>(unwind_codes.size());
+
+    new_unwind_info.frame_register_and_offset = 0x00;
+    if (options.use_frame_pointer) {
+      new_unwind_info.frame_register_and_offset =
+          PackFrameRegisterAndOffset(options.frame_pointer_register, options.frame_pointer_offset);
+    }
+
+    new_unwind_info.unwind_codes = unwind_codes;
+    new_unwind_info.exception_handler_address = 0L;
+
+    // TODO: Chained info.
+    new_unwind_info.chained_info = RuntimeFunction{0, 0, 0};
+
+    *unwind_info = new_unwind_info;
+  }
+
+ protected:
+  PeCoffUnwindInfoUnwinderX86_64 unwinder_;
+  std::unique_ptr<MemoryFake> process_mem_fake_;
+
+  // Stack pointer, 16-byte aligned, starting with a large value as we grow downwards
+  uint64_t stack_pointer_ = std::numeric_limits<uint64_t>::max() & ~0x0F;
+};
+
+TEST_F(PeCoffUnwindInfoUnwinderX86_64Test, eval_fails_inconsistent_num_codes) {
+  UnwindInfo unwind_info;
+  unwind_info.num_codes = 2;
+
+  RegsX86_64 regs;
+  constexpr uint64_t kCodeOffset = 0x0;
+  uint64_t frame_pointer;
+  bool frame_pointer_used;
+
+  EXPECT_FALSE(unwinder_.Eval(process_mem_fake_.get(), &regs, unwind_info, kCodeOffset,
+                              &frame_pointer, &frame_pointer_used));
+}
+
+TEST_F(PeCoffUnwindInfoUnwinderX86_64Test, eval_fails_with_unsupported_machframe_instruction) {
+  UnwindCode unwind_code;
+  unwind_code.code_and_op.code_offset = 1;
+  unwind_code.code_and_op.unwind_op_and_op_info = PackUnwindOpAndOpInfo(UWOP_PUSH_MACHFRAME, 0x00);
+
+  UnwindInfo unwind_info;
+  unwind_info.unwind_codes.emplace_back(unwind_code);
+  unwind_info.num_codes = unwind_info.unwind_codes.size();
+
+  RegsX86_64 regs;
+  constexpr uint64_t kCodeOffset = 0x0;
+  uint64_t frame_pointer;
+  bool frame_pointer_used;
+
+  EXPECT_FALSE(unwinder_.Eval(process_mem_fake_.get(), &regs, unwind_info, kCodeOffset,
+                              &frame_pointer, &frame_pointer_used));
```

Comment:
What about checking the error code here - similar to the other tests?

Reply:
Yes, thanks for spotting this, done.